Issue 87

EEM86 opened the pull request:
The code seems to be too coupled. It will be hard to extend this code when you add new distance items. 
Also, I refactored conversion formulas. It is better to create a new class for that, but for now, I think it is ok to have public vars in the interface.
And the most important - using float and double for formulas is not a good idea. The calculations aren't quite correct.

In the scope of this task seem all is done.

Commit:
issue-87: add conversion for distance units

Segelzwerg commented on src/main/java/segelzwerg/sporttooolbox/iunits/distance/Nautical.java:
```diff
@@ -86,6 +84,16 @@ public Kilometer toKilometer() {
 
+    @Override
+    public Miles toMiles() {
+        return new Miles(getNautical() * NAUTICAL_TO_KM / MILES_TO_KM);
```
```suggestion
        return new Miles(getNautical() * NAUTICAL_TO_MILES);
```
shouldn't we use the one constant instead of two here?

Segelzwerg commented on src/main/java/segelzwerg/sporttooolbox/iunits/distance/Distance.java:
```diff
@@ -10,7 +10,11 @@
 public interface Distance {
     double MILES_TO_YARDS = 1760.0;
     double FATHOMS_TO_NAUTICAL_MILES = 1 / 1013.3333334;
+    float MILES_TO_KM = 1.6093444f;
+    float KILOMETER_TO_METERS = 1000f;
 
+    float NAUTICAL_TO_KM = 1.852f;
+    float NAUTICAL_TO_FATHOMS = 1012.685914261f;
```
I think it's better to have these variables inside the classes because `Miles` don't need to know about `Kilometer's` conversion.

EEM86 replied:
Yeah, I moved them for an example in this task. As for me, the better solution is to have a service that would convert distance instances to each other. I think I'll show you my suggestions for the code this week.

Segelzwerg reviewed (changes requested):
> The code seems to be too coupled. It will be hard to extend this code when you add new distance items.
Well yes, for new units we would have to implement all conversion methods, but therefore we would have other benefits when calling methods from the outside.
> Also, I refactored conversion formulas. It is better to create a new class for that, but for now, I think it is ok to have public vars in the interface.
Again I think the benefit it is, from the outside all distance units looks the same.
> And the most important - using float and double for formulas is not a good idea. The calculations aren't quite correct.

Yes, that's correct, but the decimals required are too few to have these kinds of problems.
> In the scope of this task seem all is done.

So, if you feel I'm mistaken here, can you elaborate on why changes would be a good idea?

Commit:
Merge branch 'develop' into issue-87

Commit:
Merge branch 'develop' into issue-87

Commit:
issue-87: add distance converters and service.

In this scope, I've changed float vars to double according to best practices.
Also, changed some getters to public for getting a chance to retrieve them.

Commit:
issue-87: refactor and fix tests

EEM86 commented:
Here are my suggestions to get the code more uncoupled.
1. After added converters and converter service new possible distance objects could use with ease convert methods and no need to create another method in Interface and in its implementations. Also, I believe in a sense of single responsibility convert methods should be managed with convert service.

2. I know why getters of Distance instances were private, but I suggest making them public to get the possibility for retrieving them from other objects.

3. I changed float types of variables to double as I believe it correlates with best practices due to float restrictions.

4. Still not sure that calculates are correct. Please, check different values in tests and create a task if you agree with me.

Will be waiting for your opinion on the mentioned suggestions.

Segelzwerg commented:
> Here are my suggestions to get the code more uncoupled.
> 
> 1. After added converters and converter service new possible distance objects could use with ease convert methods and no need to create another method in Interface and in its implementations. Also, I believe in a sense of single responsibility convert methods should be managed with convert service.
> 2. I know why getters of Distance instances were private, but I suggest making them public to get the possibility for retrieving them from other objects.
> 3. I changed float types of variables to double as I believe it correlates with best practices due to float restrictions.
> 4. Still not sure that calculates are correct. Please, check different values in tests and create a task if you agree with me.
> 
> Will be waiting for your opinion on the mentioned suggestions.

Hi @EEM86  thank you for your suggestion and patience. I had a busy week. I agree with you that is kind of violating the SRP, but implementing a converter for every unit it results in the same amount of code being implemented for new units.

I discussed the idea with the colleague and he is suggesting that we should implement a SI system: https://github.com/iTitus/commons/commit/0a1828dc01f3be1ee5f262100c9b346b77080b30
I kind of like that idea as it would remove a lot of constraints and converting units we basically by its self. So maybe we should wait a bit and see if the idea of @iTitus would be an improvement.

iTitus commented:
I got most of it set up now over at https://github.com/iTitus/commons/tree/master/src/main/java/io/github/ititus/si
I will add a few missing units and maybe tweak the value type and then prototype a PR.
